src/commands/build/templatedatavalidator.js

```
@@ -0,0 +1,41 @@
+const fs = require('file-system');
+const UserError = require('../../errors/usererror');
+const { isCustomError } = require('../../utils/errorutils');
+const { info } = require('../../utils/logger');
+
+/**
+ * TemplateDataValidator is reponsible for checking data supplied to a page
+ * using Theme's custom validation steps (if any).
+ */
+module.exports = class TemplateDataValidator {
+  constructor(templateDataValidationHook) {
+    /**
+     * The path to template data validation hook.
+     * @type {string}
+     */
+    this._templateDataValidationHook = templateDataValidationHook;
+  }
+  
+  /**
+   * Execute validation hook's function if file exists
+   * @param {string} pageName name of the current page
+   * @param {Object} pageData template arguments for the current page
+   */
+  validate({ pageName, pageData }) {
+    if (!fs.existsSync(this._templateDataValidationHook)) {
+        return;
+    }
+    try {
+        info(`Validating configuration for page "${pageName}".`);
+        const validatorFunction = require(this._templateDataValidationHook);
+        validatorFunction(pageData);
```

I'm wondering if we want this class and the hook to throw an exception when the data is invalid. The alternative is that both return a boolean. I think the boolean has its advantages:
- Consumers of the validators can more easily decide how they want to handle bad templates. It may be that they thrown an exception if the validation result is `false`. Or, they try and handle more gracefully. As an example, in the future, maybe users can configure Jambo to gracefully ignore bad pages and continue generation (although I think the more catastrophic failure is the way to go for now). 
- We can differentiate between an actual issue running the Theme supplied validator and an invalid template. Right now, if there was an exception thrown because the validator couldn't run properly, we would treat that the same as the validator finding the template to be no good. We should probably treat the two situations differently and give different errors.

---

That makes sense, it would give user more flexibility on exception handling. I'm thinking of having the hook returns a boolean, like you suggested, and use that for the validator class to determines whether it should throw the exception or not. For the second point, I believe you can differentiate the two issues right now. Assuming the theme would provide a custom error (UserError) for any template issues, the validator would throw that accordingly. Otherwise, it must be something wrong with running the hook itself so validator should throw a new error with a fail to execute message.

---

That makes sense, it would give user more flexibility on exception handling. I'm thinking of having the hook returns a boolean, like you suggested, and have this class use the boolean to determines whether or not it should throw an exception. This would fall on the Theme’s hook to log in detail what is invalid, and the validator will throw a new error of general bad template if hook return true.